### PR TITLE
Fix AI spamming harvester resource searches

### LINKED ISSUE #14467: AI considers its harvesters 'idle' way too often
In short, the check for idle harvesters in HackyAI is flat out crap, making the AI perform expensive, map-wide resource searches far too often.

### PR DESCRIPTION
**Pull request by reaperrr**
Fixes #14467.

### CODE REVIEW

**Review by MustaphaTR — APPROVED** [review]
Looks good to me. 👍

**Comment by reaperrr** [comment]
Updated, now also covers the case @airetaM pointed out in #14467.
The AI now checks if there's even a ResourceLayer - and if yes, whether there are any cells with resources - before giving orders to harvesters.

The difference on maps without resources is outright massive. I'd actually like to see this in the playtest.

**Review by Smittytron — APPROVED** [review]
Tested this vs current release on map without resources. Easy to see the performance improvement. 👍

**Review by MustaphaTR — CHANGES_REQUESTED** [review]
Something is off with 3rd commit. AI is no longer doing anything with harvs.

I have something like that in a testmap, also a 3x3 patch at other side of the map, after here finishes AI no longer searches for other one. 
resim

Similarly ones it built from WFac don't go to harvesting there, as FindResources can't find the other ore on the map.

File: OpenRA.Mods.Common/Traits/World/ResourceLayer.cs [hunk 1]
```diff
@@ -251,6 +253,8 @@ public ResourceType Harvest(CPos cell)
 			if (c.Type == null)
 				return null;
 
+			--resCells;
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/World/ResourceLayer.cs]: Shouldn't this be inside the `if` below?

Comment (MustaphaTR) [reply OpenRA.Mods.Common/Traits/World/ResourceLayer.cs]: Looks like indeed that's the reason. My testcase works fine with this moved there.

**Comment by reaperrr** [comment]
Updated.

Also added a fix for https://github.com/OpenRA/OpenRA/issues/14467#issuecomment-349967217 (at some point the `IsInAir` boolean might come in handy in other places, too).

File: OpenRA.Mods.Common/Traits/World/ResourceLayer.cs [hunk 2]
```diff
@@ -288,6 +293,11 @@ public int GetMaxResourceDensity(CPos cell)
 			return Content[cell].Type.Info.MaxDensity;
 		}
 
+		public bool IsResourceLayerEmpty()
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/World/ResourceLayer.cs]: Nit: this would be better as a property than a method.

**Review by pchote — DISMISSED** [review]
Works as advertised :+1:

Just one minor code nit:

**Comment by reaperrr** [comment]
Updated.

**Review by pchote — APPROVED** [review]

**Comment by abcdefg30** [comment]
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/9f74c1a4b45270f22e43aa896ce9415033cf4871...05866958643bd1f25f93142aa4b8eba3ea95b396)

**Pull request merged**